client/src/details/UrlBuilder.cpp

```diff
@@ -66,42 +93,82 @@ std::string UrlBuilder::GetUrl() const
     return urlPtr;
 }
 
-void UrlBuilder::SetScheme(Scheme scheme)
+UrlBuilder& UrlBuilder::SetScheme(Scheme scheme)
 {
     switch (scheme)
     {
     case Scheme::Https:
         THROW_IF_CURL_URL_SETUP_ERROR(curl_url_set(m_handle, CURLUPART_SCHEME, "https", 0 /*flags*/));
         break;
     }
+    return *this;
 }
 
-void UrlBuilder::SetHost(const std::string& host)
+UrlBuilder& UrlBuilder::SetHost(const std::string& host)
 {
     THROW_IF_CURL_URL_SETUP_ERROR(curl_url_set(m_handle, CURLUPART_HOST, host.c_str(), 0 /*flags*/));
+    return *this;
 }
 
-void UrlBuilder::SetPath(const std::string& path, bool encode)
+UrlBuilder& UrlBuilder::SetPath(const std::string& path, bool encode)
 {
     unsigned flags = 0;
     if (encode)
     {
         flags |= CURLU_URLENCODE;
     }
     THROW_IF_CURL_URL_SETUP_ERROR(curl_url_set(m_handle, CURLUPART_PATH, path.c_str(), flags));
+    return *this;
+}
+
+UrlBuilder& UrlBuilder::AppendPath(const std::string& path, bool encode)

```

Comment:
nit: inefficient?  what if we call AppendPath several times?  Each time it's calloing curl_url_get, escaping and appending?  Is it possible to keep a "raw" std::string in this class until a "final" string (e.g. via "get" accessor) is requested by caller?

Reply:
> Is it possible to keep a "raw" std::string in this class until a "final" string (e.g. via "get" accessor) is requested by caller?

That would violate the expectation that Get() is const and only returns existing data.

What we could do is keep a raw string locally that is in sync with what is in curl_url_get(CURLUPART_PATH).
That saves the repeated get call. In the end the memory consumption will be constant of two strings for URL instead of fluctuating once AppendPath ends.
Overall it would only change the performance time of the calls.